crates/client/eth/src/l1_gas_price.rs

```diff
@@ -0,0 +1,235 @@
+) -> anyhow::Result<()> {
+    let fee_history = eth_client.provider.get_fee_history(300, BlockNumberOrTag::Latest, &[]).await?;
+
+    // The RPC responds with 301 elements for some reason. It's also just safer to manually
+    // take the last 300. We choose 300 to get average gas caprice for last one hour (300 * 12 sec block
+    // time).
+    let (_, blob_fee_history_one_hour) =
+        fee_history.base_fee_per_blob_gas.split_at(fee_history.base_fee_per_blob_gas.len().max(300) - 300);
+
+    let avg_blob_base_fee = blob_fee_history_one_hour.iter().sum::<u128>() / blob_fee_history_one_hour.len() as u128;
+
+    let eth_gas_price = fee_history.base_fee_per_blob_gas.last().context("Setting l1 last confirmed block number")?;
+
+    l1_data_provider.update_eth_l1_gas_price(*eth_gas_price).await;
+    l1_data_provider.update_eth_l1_data_gas_price(avg_blob_base_fee).await;
+
+    update_last_update_timestamp().await;
+
+    // Update block number separately to avoid holding the lock for too long
+    update_l1_block_metrics(eth_client, l1_data_provider).await?;
+
+    Ok(())
+}
+
+async fn update_l1_block_metrics(
+    eth_client: &EthereumClient,
+    l1_data_provider: Arc<dyn L1DataProvider>,
+) -> anyhow::Result<()> {
+    // Get the latest block number
+    let latest_block_number = eth_client.get_latest_block_number().await?;
+
+    // Get the current gas price
+    let current_gas_price = l1_data_provider.get_gas_prices().await;
+    let eth_gas_price = current_gas_price.eth_l1_gas_price;
+
+    // Update the metrics
+    eth_client.l1_block_metrics.l1_block_number.set(latest_block_number as f64);
+    eth_client.l1_block_metrics.l1_gas_price_wei.set(eth_gas_price as f64);
+
+    // We're ignoring l1_gas_price_strk
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod eth_client_gas_price_worker_test {
+    use super::*;
+    use crate::client::eth_client_getter_test::create_ethereum_client;
+    use dc_mempool::GasPriceProvider;
+    use futures::future::FutureExt;
+    use httpmock::{MockServer, Regex};
+    use rstest::*;
+    use std::panic::AssertUnwindSafe;
+    use std::time::{SystemTime, UNIX_EPOCH};
+    use tokio::time::{timeout, Duration};
+
+    #[fixture]
+    #[once]
+    pub fn eth_client_with_mock() -> (MockServer, EthereumClient) {
+        let server = MockServer::start();
+        let addr = format!("http://{}", server.address());
+        let eth_client = create_ethereum_client(Some(&addr));
+        (server, eth_client)
+    }
+
+    #[fixture]
+    #[once]
+    pub fn eth_client() -> EthereumClient {
+        create_ethereum_client(None)
+    }
+
+    #[rstest]
+    #[tokio::test]
+    async fn gas_price_worker_works(eth_client: &'static EthereumClient) {
+        let l1_data_provider: Arc<dyn L1DataProvider> = Arc::new(GasPriceProvider::new());
+
+        // Run the worker for a short time
+        let worker_handle = tokio::spawn(gas_price_worker(eth_client, l1_data_provider.clone(), false));
```

Comment:
- if `infinite_loop` is false, do we need to spawn tokio? we can just await here right?
- we should also write a case for infinite loop true
- how does the current code work without a delay?

Reply:
- yeah I think that should work, using the await only
- will do, we have that when we are testing for the fail case but will create for this as well
- there is a delay, it's in the eth_client